sycl/include/sycl/ext/intel/experimental/device_architecture.hpp

```diff
@@ -0,0 +1,301 @@
+#endif
+#ifndef __SYCL_TARGET_INTEL_GPU_ICLLP__
+#define __SYCL_TARGET_INTEL_GPU_ICLLP__ 0
+#endif
+#ifndef __SYCL_TARGET_INTEL_GPU_EHL__
+#define __SYCL_TARGET_INTEL_GPU_EHL__ 0
+#endif
+#ifndef __SYCL_TARGET_INTEL_GPU_TGLLP__
+#define __SYCL_TARGET_INTEL_GPU_TGLLP__ 0
+#endif
+#ifndef __SYCL_TARGET_INTEL_GPU_RKL__
+#define __SYCL_TARGET_INTEL_GPU_RKL__ 0
+#endif
+#ifndef __SYCL_TARGET_INTEL_GPU_ADL_S__
+#define __SYCL_TARGET_INTEL_GPU_ADL_S__ 0
+#endif
+#ifndef __SYCL_TARGET_INTEL_GPU_RPL_S__
+#define __SYCL_TARGET_INTEL_GPU_RPL_S__ 0
+#endif
+#ifndef __SYCL_TARGET_INTEL_GPU_ADL_P__
+#define __SYCL_TARGET_INTEL_GPU_ADL_P__ 0
+#endif
+#ifndef __SYCL_TARGET_INTEL_GPU_ADL_N__
+#define __SYCL_TARGET_INTEL_GPU_ADL_N__ 0
+#endif
+#ifndef __SYCL_TARGET_INTEL_GPU_DG1__
+#define __SYCL_TARGET_INTEL_GPU_DG1__ 0
+#endif
+#ifndef __SYCL_TARGET_INTEL_GPU_ACM_G10__
+#define __SYCL_TARGET_INTEL_GPU_ACM_G10__ 0
+#endif
+#ifndef __SYCL_TARGET_INTEL_GPU_ACM_G11__
+#define __SYCL_TARGET_INTEL_GPU_ACM_G11__ 0
+#endif
+#ifndef __SYCL_TARGET_INTEL_GPU_ACM_G12__
+#define __SYCL_TARGET_INTEL_GPU_ACM_G12__ 0
+#endif
+#ifndef __SYCL_TARGET_INTEL_GPU_PVC__
+#define __SYCL_TARGET_INTEL_GPU_PVC__ 0
+#endif
+
+// This is true when the translation unit is compiled in AOT mode with target
+// names that supports the "if_architecture_is" features.  If an unsupported
+// target name is specified via "-fsycl-targets", the associated invocation of
+// the device compiler will set this variable to false, and that will trigger
+// an error for code that uses "if_architecture_is".
+static constexpr bool is_allowable_aot_mode =
+    (__SYCL_TARGET_INTEL_X86_64__ == 1) ||
+    (__SYCL_TARGET_INTEL_GPU_BDW__ == 1) ||
+    (__SYCL_TARGET_INTEL_GPU_SKL__ == 1) ||
+    (__SYCL_TARGET_INTEL_GPU_KBL__ == 1) ||
+    (__SYCL_TARGET_INTEL_GPU_CFL__ == 1) ||
+    (__SYCL_TARGET_INTEL_GPU_APL__ == 1) ||
+    (__SYCL_TARGET_INTEL_GPU_GLK__ == 1) ||
+    (__SYCL_TARGET_INTEL_GPU_WHL__ == 1) ||
+    (__SYCL_TARGET_INTEL_GPU_AML__ == 1) ||
+    (__SYCL_TARGET_INTEL_GPU_CML__ == 1) ||
+    (__SYCL_TARGET_INTEL_GPU_ICLLP__ == 1) ||
+    (__SYCL_TARGET_INTEL_GPU_EHL__ == 1) ||
+    (__SYCL_TARGET_INTEL_GPU_TGLLP__ == 1) ||
+    (__SYCL_TARGET_INTEL_GPU_RKL__ == 1) ||
+    (__SYCL_TARGET_INTEL_GPU_ADL_S__ == 1) ||
+    (__SYCL_TARGET_INTEL_GPU_RPL_S__ == 1) ||
+    (__SYCL_TARGET_INTEL_GPU_ADL_P__ == 1) ||
+    (__SYCL_TARGET_INTEL_GPU_ADL_N__ == 1) ||
+    (__SYCL_TARGET_INTEL_GPU_DG1__ == 1) ||
+    (__SYCL_TARGET_INTEL_GPU_ACM_G10__ == 1) ||
+    (__SYCL_TARGET_INTEL_GPU_ACM_G11__ == 1) ||
+    (__SYCL_TARGET_INTEL_GPU_ACM_G12__ == 1) ||
+    (__SYCL_TARGET_INTEL_GPU_PVC__ == 1);
+
+struct IsAOTForArchitectureClass {
+  // allocate an array of size == size of ext::intel::experimental::architecture
+  // enum
+  bool arr[static_cast<int>(max_architecture) + 1];
+
+  constexpr IsAOTForArchitectureClass() : arr() {
+    arr[static_cast<int>(ext::intel::experimental::architecture::x86_64)] =
```

Comment:
At the very least, can we add `using arch = ext::intel::experimental::architecture` here? It would be local to this function and wouldn't escape it.

I think we can further simplify it by using some constant "mapping" and iterating over it, but that would likely be an overkill.

On the other hand, can we use the `#include <name.inc>` trick to make it more maintainable maybe? And if we do that, then the `max_architecture` might be computed in compile time automatically by doing `0 +1 +1 +1` using that same `<name.inc>`.

Reply:
>> can we add using arch = ext::intel::experimental::architecture here
Done.

>> On the other hand, can we use the #include <name.inc> trick t
Can be applied in separate PR.